File: python/golf/reader.py
"""GolfReader: read-only access to .golf files with range query support."""

import mmap
import struct
from pathlib import Path
from typing import BinaryIO, List, Optional, Union

from golf.core import (
    BLOCK_DESCRIPTOR_SIZE,
    FIXED_HEADER_SIZE,
    FOOTER_SIZE,
    BlockDescriptor,
    Footer,
    Header,
    Record,
    compute_crc32c,
    decode_metadata,
    decompress_block,
    find_first_block,
    find_last_block,
    MetadataEntry,
)


def _binary_search_first(data: bytes, record_size: int, count: int, target: int) -> int:
    """Find first record index where timestamp >= target."""
    lo, hi = 0, count
    while lo < hi:
        mid = lo + (hi - lo) // 2
        ts = struct.unpack_from("<Q", data, mid * record_size)[0]
        if ts < target:
            lo = mid + 1
        else:
            hi = mid
    return lo


def _binary_search_last(data: bytes, record_size: int, count: int, target: int) -> int:
    """Find last record index where timestamp <= target."""
    lo, hi = 0, count
    while lo < hi:
        mid = lo + (hi - lo) // 2
        ts = struct.unpack_from("<Q", data, mid * record_size)[0]
        if ts <= target:
            lo = mid + 1
        else:
            hi = mid
    return max(0, lo - 1)


def _collect_range(
    descriptors: List[BlockDescriptor],
    compression,
    record_value_size: int,
    start_ts: int,
    end_ts: int,
    read_block,
) -> List[Record]:
    """Shared range query logic."""
    first_idx = find_first_block(descriptors, start_ts)
    if first_idx < 0:
        return []
    last_idx = find_last_block(descriptors, end_ts)
    if last_idx < 0:
        return []
    if first_idx > last_idx:
        return []

    record_size = 8 + record_value_size
    results = []

    for bi in range(first_idx, last_idx + 1):
        desc = descriptors[bi]
        compressed = read_block(desc)

        raw = decompress_block(compressed, compression, desc.uncompressed_size)

        actual_crc = compute_crc32c(raw)
        if actual_crc != desc.block_crc:
            raise ValueError(
                f"block CRC mismatch: expected {desc.block_crc:#010x}, "
                f"got {actual_crc:#010x}"
            )

        count = desc.record_count
        rec_start = 0
        rec_end = count - 1

        if bi == first_idx:
            rec_start = _binary_search_first(raw, record_size, count, start_ts)
        if bi == last_idx:
            rec_end = _binary_search_last(raw, record_size, count, end_ts)

        if rec_start <= rec_end < count:
            for i in range(rec_start, rec_end + 1):
                off = i * record_size
                ts = struct.unpack_from("<Q", raw, off)[0]
                if start_ts <= ts <= end_ts:
                    val = raw[off + 8 : off + record_size]
                    results.append(Record(timestamp=ts, value=bytes(val)))

    return results


class GolfReader:
    """Reads .golf files using standard I/O (read + seek)."""

    def __init__(self, f: BinaryIO):
        self._f = f

        # Get file size
        f.seek(0, 2)
        file_size = f.tell()
        if file_size < FIXED_HEADER_SIZE + FOOTER_SIZE:
            raise ValueError(f"file too small: {file_size} bytes")

        # Read footer
        f.seek(-FOOTER_SIZE, 2)
        footer_buf = f.read(FOOTER_SIZE)
        footer = Footer.decode(footer_buf)

        # Read block index
        index_size = footer.block_count * BLOCK_DESCRIPTOR_SIZE
        f.seek(footer.index_offset)
        index_buf = f.read(index_size)
        actual_crc = compute_crc32c(index_buf)
        if actual_crc != footer.index_crc:
            raise ValueError("block index CRC mismatch")

        self.descriptors: List[BlockDescriptor] = []
        for i in range(footer.block_count):
            start = i * BLOCK_DESCRIPTOR_SIZE
            chunk = index_buf[start : start + BLOCK_DESCRIPTOR_SIZE]
            self.descriptors.append(BlockDescriptor.decode(chunk))

        # Read header
        f.seek(0)
        header_buf = f.read(FIXED_HEADER_SIZE)
        self.header = Header.decode(header_buf)

        # Read metadata
        self.metadata: List[MetadataEntry] = []
        if self.header.metadata_length > 0:
            meta_buf = f.read(self.header.metadata_length)
            self.metadata = decode_metadata(meta_buf)

    @classmethod
    def open(cls, path: Union[str, Path]) -> "GolfReader":
        """Open a golf file by path."""
        f = open(path, "rb")
        return cls(f)

    @property
    def record_count(self) -> int:
        return self.header.record_count

    @property
    def timestamp_range(self):
        return (self.header.min_timestamp, self.header.max_timestamp)

    def query(self, start_ts: int, end_ts: int) -> List[Record]:
        """Query records in the inclusive range [start_ts, end_ts]."""

        def read_block(desc: BlockDescriptor) -> bytes:
            self._f.seek(desc.block_offset)
            return self._f.read(desc.compressed_size)

        return _collect_range(
            self.descriptors,
            self.header.compression,
            self.header.record_value_size,
            start_ts,
            end_ts,
            read_block,
        )

    def close(self):
        self._f.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()


class MmapGolfReader:
    """Reads .golf files using memory-mapped I/O."""

    def __init__(self, path: Union[str, Path]):
        self._file = open(path, "rb")
        self._mmap = mmap.mmap(self._file.fileno(), 0, access=mmap.ACCESS_READ)
        data = self._mmap

        size = len(data)
        if size < FIXED_HEADER_SIZE + FOOTER_SIZE:
            raise ValueError(f"file too small: {size} bytes")

        # Parse footer
        footer_buf = data[size - FOOTER_SIZE : size]
        footer = Footer.decode(bytes(footer_buf))

        # Parse block index
        idx_start = footer.index_offset
        idx_end = idx_start + footer.block_count * BLOCK_DESCRIPTOR_SIZE
        index_buf = bytes(data[idx_start:idx_end])
        actual_crc = compute_crc32c(index_buf)
        if actual_crc != footer.index_crc:
            raise ValueError("block index CRC mismatch")

        self.descriptors: List[BlockDescriptor] = []
        for i in range(footer.block_count):
            start = i * BLOCK_DESCRIPTOR_SIZE
            chunk = index_buf[start : start + BLOCK_DESCRIPTOR_SIZE]
            self.descriptors.append(BlockDescriptor.decode(chunk))

        # Parse header
        header_buf = bytes(data[0:FIXED_HEADER_SIZE])
        self.header = Header.decode(header_buf)

        # Parse metadata
        self.metadata: List[MetadataEntry] = []
        if self.header.metadata_length > 0:
            meta_buf = bytes(
                data[FIXED_HEADER_SIZE : FIXED_HEADER_SIZE + self.header.metadata_length]
            )
            self.metadata = decode_metadata(meta_buf)

    @property
    def record_count(self) -> int:
        return self.header.record_count

    @property
    def timestamp_range(self):
        return (self.header.min_timestamp, self.header.max_timestamp)

    def query(self, start_ts: int, end_ts: int) -> List[Record]:
        """Query records in the inclusive range [start_ts, end_ts]."""
        data = self._mmap

        def read_block(desc: BlockDescriptor) -> bytes:
            return bytes(data[desc.block_offset : desc.block_offset + desc.compressed_size])

        return _collect_range(
            self.descriptors,
            self.header.compression,
            self.header.record_value_size,
            start_ts,
            end_ts,
            read_block,
        )

    def close(self):
        self._mmap.close()
        self._file.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

File: kotlin/src/main/kotlin/golf/Reader.kt
// Reader.kt — read-only access to .golf files with indexed range queries.

package golf

import java.io.File

/**
 * Read-only accessor for a golf file, loaded fully into memory.
 *
 * Opening validates the footer, block index, and header (magic + CRC32C) and
 * decodes the metadata. [query] then binary-searches the block index so only
 * blocks overlapping the requested range are touched; untouched blocks are
 * never decompressed. Files are immutable after sealing, so instances are safe
 * to share across reads.
 *
 * Per-block data integrity is verified lazily: each block's CRC32C is checked
 * as it is first decompressed during a query.
 */
class GolfReader private constructor(
    /** Validated fixed header of the file. */
    val header: GolfHeader,
    /** Decoded header metadata (empty when none was written). */
    val metadata: List<MetadataEntry>,
    /** Block-index descriptors in file order. */
    val descriptors: List<BlockDescriptor>,
    private val data: ByteArray,
) {
    private val recordSize = 8 + header.recordValueSize

    /** Total records in the file (from the validated header). */
    val recordCount: ULong get() = header.recordCount

    /** Inclusive `[min, max]` timestamp span of the file. */
    val timestampRange: ClosedRange<ULong>
        get() = header.minTimestamp..maxOf(header.minTimestamp, header.maxTimestamp)

    companion object {
        /** Loads and parses the file at [path]. */
        fun open(path: String): GolfReader = fromBytes(File(path).readBytes())

        /** Loads and parses the given [file]. */
        fun open(file: File): GolfReader = fromBytes(file.readBytes())

        /**
         * Parses an in-memory golf file image. The array is retained by the
         * reader (not copied), so callers must not mutate it afterwards.
         */
        fun fromBytes(data: ByteArray): GolfReader {
            if (data.size < FIXED_HEADER_SIZE + FOOTER_SIZE) {
                throw GolfException("file too small: ${data.size} bytes")
            }

            // Footer → block index → header; each step is checksum-validated.
            // Offset arithmetic runs in the unsigned domain with bounds
            // checked *before* any narrowing, so crafted u64s surface as
            // typed errors instead of truncation or OutOfMemoryError.
            val footer = Footer.decode(data.copyOfRange(data.size - FOOTER_SIZE, data.size))
            if (footer.indexOffset > data.size.toULong() ||
                footer.blockCount > (data.size / BLOCK_DESCRIPTOR_SIZE).toULong()
            ) {
                throw GolfException(
                    "corrupted index: offset ${footer.indexOffset}, ${footer.blockCount} blocks outside file",
                )
            }
            val indexStart = footer.indexOffset.toInt()
            val indexEnd = indexStart + footer.blockCount.toInt() * BLOCK_DESCRIPTOR_SIZE // bounded ⇒ no overflow
            if (indexEnd > data.size - FOOTER_SIZE) {
                throw GolfException(
                    "corrupted index: range $indexStart..$indexEnd outside file",
                )
            }
            val indexBytes = data.copyOfRange(indexStart, indexEnd)
            if (Crc32c.checksum(indexBytes) != footer.indexCrc) {
                throw GolfException("block index CRC mismatch")
            }
            // Capacity hint is clamped: a hostile blockCount can't make the
            // ArrayList preallocation itself fail.
            val descriptors = ArrayList<BlockDescriptor>(minOf(footer.blockCount.toInt(), 4096))
            repeat(footer.blockCount.toInt()) { i ->
                descriptors.add(BlockDescriptor.decode(indexBytes, i * BLOCK_DESCRIPTOR_SIZE))
            }

            val header = GolfHeader.decode(data.copyOf(0.coerceAtLeast(FIXED_HEADER_SIZE)))

            val metadata = if (header.metadataLength > 0) {
                val end = FIXED_HEADER_SIZE + header.metadataLength
                if (end > data.size) throw GolfException("metadata length exceeds file")
                decodeMetadata(data.copyOfRange(FIXED_HEADER_SIZE, end))
            } else emptyList()

            return GolfReader(header, metadata, descriptors, data)
        }
    }

    /**
     * Returns every record with `startTs <= timestamp <= endTs`, ascending.
     *
     * Only blocks overlapping the range are read and decompressed; within each
     * boundary block the scan is narrowed by timestamp binary search.
     */
    fun query(startTs: ULong, endTs: ULong): List<Record> {
        if (descriptors.isEmpty()) return emptyList()
        val firstIdx = findFirstBlock(descriptors, startTs)
        val lastIdx = findLastBlock(descriptors, endTs)
        if (firstIdx < 0 || lastIdx < 0 || firstIdx > lastIdx) return emptyList()

        val results = ArrayList<Record>()
        for (bi in firstIdx..lastIdx) {
            val desc = descriptors[bi]
            val raw = decompressedBlock(desc)

            val recStart = if (bi == firstIdx) {
                binarySearchFirst(raw, recordSize, desc.recordCount, startTs)
            } else 0
            val recEnd = if (bi == lastIdx) {
                binarySearchLast(raw, recordSize, desc.recordCount, endTs)
            } else desc.recordCount - 1

            if (recStart > recEnd) continue
            for (i in recStart..recEnd) {
                val ts = readTimestamp(raw, i, recordSize)
                if (ts >= startTs && ts <= endTs) {
                    results.add(extractRecord(raw, i, header.recordValueSize))
                }
            }
        }
        return results
    }

    /** Returns every record whose timestamp falls inside [range], ascending. */
    fun query(range: ClosedRange<ULong>): List<Record> = query(range.start, range.endInclusive)

    /** Loads the block described by [desc] and verifies its CRC32C. */
    private fun decompressedBlock(desc: BlockDescriptor): ByteArray {
        // Narrow only after unsigned-domain bounds checks.
        if (desc.blockOffset > data.size.toULong() ||
            desc.compressedSize.toLong() !in 0..data.size.toLong()
        ) {
            throw GolfException("corrupted block: offset ${desc.blockOffset}, size ${desc.compressedSize} outside file")
        }
        val start = desc.blockOffset.toInt()
        val end = start + desc.compressedSize
        if (end < start || end > data.size) {
            throw GolfException("corrupted block: range $start..$end outside file")
        }
        val payload = data.copyOfRange(start, end)

        val raw: ByteArray = when (header.compression) {
            Compression.NONE -> payload
            Compression.LZ4 -> {
                if (payload.size < 4) throw GolfException("LZ4 data too short")
                val origSize = Le.u32(payload, 0).toInt()
                Lz4Block.decompress(payload, 4, payload.size - 4, origSize)
            }
            Compression.ZSTD ->
                throw GolfException("Zstd reading is not supported by this implementation")
        }

        if (Crc32c.checksum(raw) != desc.blockCrc) {
            throw GolfException("block CRC mismatch")
        }
        return raw
    }
}
